### PR TITLE
db: _init: only try creating missing subdirs

### PR DESCRIPTION
**Pull request by efiop**
As pointed out by @skshetry before in https://github.com/iterative/dvc-objects/pull/147#issuecomment-1249362678 we can at least try to `ls` to avoid redundant api calls.

Relevant for slow NFS setups as pointed out by @dberenbaum during his experiments on his home NAS.

### CODE REVIEW

**Comment by skshetry** [comment]
This has broken `dvc-http` since there is no `ls()` in HTTPFileSystem.
See https://github.com/iterative/dvc-objects/pull/163. As a result, `2.33.0` is broken.
Related: https://github.com/iterative/dvc/pull/8487.

File: src/dvc_objects/db.py [hunk 1]
```diff
@@ -50,8 +51,16 @@ def _init(self):
         if self._initialized:
             return
 
-        for num in range(0, 256):
-            self.makedirs(self.fs.path.join(self.path, f"{num:02x}"))
+        dnames = {f"{num:02x}" for num in range(0, 256)}
+        existing = set()
+        with suppress(FileNotFoundError):
```
Comment (skshetry) [inline src/dvc_objects/db.py:56]: Should probably also catch `NotImplementedError`.

File: src/dvc_objects/db.py [hunk 2]
```diff
@@ -50,8 +51,16 @@ def _init(self):
...
+            existing = {
+                self.fs.path.basename(path)
+                for path in self.fs.ls(self.path, detail=False)
+            }
+
+        for dname in dnames.difference(existing):
+            self.makedirs(self.fs.path.join(self.path, dname))
```
Comment (skshetry) [inline src/dvc_objects/db.py:63]: I meant this to be lazy actually.

File: src/dvc_objects/db.py (hunk 1 above)
Comment (skshetry) [reply src/dvc_objects/db.py:56]: And other `OSError`s too?

Comment (efiop) [reply src/dvc_objects/db.py:56]: Ah, right, for http. Can see a test failure now too.

**Pull request merged**